packages/backend-core/src/events/processors/DocumentUpdateProcessor.ts

```diff
@@ -0,0 +1,42 @@
+import { Event, Identity, Group, DocumentUpdateEvents } from "@budibase/types"
+import { EventProcessor } from "./types"
+import * as docUpdates from "../../docUpdates"
+import { getTenantId } from "../../context"
+
+export default class DocumentUpdateProcessor implements EventProcessor {
```

Comment:
I think we can do a bit of refactor around the difference between event publisher and event processor. 

Here we are using an event _processor_ as a _publisher_, with a separate concept of processing inside `docUpdates/updates`.

Ideally we would have a unified interface around event processing and publishing and slot the document processing into that. 

For example we can move publishing out of `DocumentUpdateProcessor` and place it here instead:
https://github.com/Budibase/budibase/blob/develop/packages/backend-core/src/events/events.ts#L6

We could still keep the `DocumentUpdateEvents` as a whitelist (but maybe change it to `AsyncEvents`) for when to do asynchronous job based processing (although ultimately I think every event should be processed this way). Could then use it like 

```ts
interface EventPayload {
    event: Event,
    identity: Identity,
    properties: any,
    timestamp?: string | number
}

if (AsyncEvents.indexOf(event) !== -1) {
   // async - as fast as possible
   const data: EventPayload = { event, identity, properties, timestamp }
   queue.add(data)
} else {
   // sync - slows down response
   processors.processEvent(...)
}
```

Reply:
See below comment about integration on the read side

Reply:
I understand the logic of this - but after some thought I don't think its a great idea to move to a half an half model for this. I'm going to refactor this a bit to make the async processing possible, using a standard event payload, but I don't like the idea of cutting over some events to async and some not.

Personally I believe this is quite outside the scope of this issue, as theres quite a risk of breaking the whole event processing pipeline by adding Bull inbetween it.

The goal of using the queue was to disconnect the producer of the event from the consumer, and allow an intensive process to become a background one, rather than re-building the event system.

I've updated it so that the `publishEvent` now fires off onto an async events queue, with a generic event payload, but it will also continue and finish the actual synchronous event processing.

I fear that disconnecting the producer of the event to the actual Posthog updates and all other parts of the event processing pipeline will come with a lot of risks - something I'd prefer to avoid when it doesn't appear to be causing any major issues.

We can open a separate issue if we want to move completely to a async model for events, with the testing required to see what impact this will have to our pipeline; where I'll leave it will be easily converted to that if desired.

Reply:
Maybe I wasn't clear, I wasn't suggesting that we cutover to async processing for everything right away - that's definitely another piece of work. What you've done around integrating into the publisher and having a generic way to init and extend events in server is exactly what I was looking for. Thanks for making the change

Reply:
I see now why the above might have been misleading, it should have been an `if` only and not an `if-else`

Reply:
This makes a lot more sense now!